[O-Darren] iP

O-Darren opened the pull request:
#Testament
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Testament is your own personal assistant. With him, you can free your mind of having to remember things you need to do! He's
- Text-based
- Easy to learn
- Provides convenience
- Entertaining!

Want your own Testament today? Simply
1. Download it [here](https://github.com/O-Darren/ip)
2. Open the jar file
3. Add your tasks
and Testament will manage your tasks for you, **free of charge** 😄 

Current features
- [x] Talks to you!
- [x] Takes in tasks
- [ ] Reminds you about deadlines (coming soon)

Want to contribute to the project? Testament is made in Java. Here is the `main` method!
```java
public static void main(String[] args) {
    new Testament("Memory.TaskList.txt").run();
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

O-Darren committed:
Rename main class

O-Darren committed:
Add Greeting

O-Darren committed:
Adjust spacing in dialogue

O-Darren committed:
Add support for echo

Use a scanner object to take in inputs, and a while loop to repeat those inputs until the exit command is given

O-Darren committed:
implement printDialogue function

write a basic function that prints a string encompassed by two lines

O-Darren committed:
Change dialogue

O-Darren committed:
Implement Task Class

Implement basic task class that takes in a String as its taskname, and returns its taskname for its toString

O-Darren committed:
Implement TaskList

Implement basic tasklist class that can take in and store tasks using an ArrayList

O-Darren committed:
TaskList.java: change add function to take in string

O-Darren committed:
TaskList.java: change toString

change toString to include task numbers

O-Darren committed:
Testament.java: implement add, list

implement add and list function through the use of Task.java and TaskList.java

O-Darren committed:
Task.java: Implement done task

Use a boolean value to track whether the task has been completed

O-Darren committed:
TaskList.java: Implement mark and unmark

Implement simple functions that mark or unmark the task at the specified index, returning true if it was successfully marked and returning false if that task does not exist

O-Darren committed:
TaskList.java: Implement getTask

implement a function that retrieves the toString representation of the specified task

O-Darren committed:
Testament.java: Implement mark and unmark tasks

O-Darren committed:
Implement ToDo class

O-Darren committed:
TaskList.java: modify add function

add function now takes in a task instead of a string

O-Darren committed:
Testament.java: add support for ToDo

O-Darren committed:
Implement deadline class

O-Darren committed:
TaskList.java: Implement getSize() function

create a simple function that retrieves the size of the tasklist

O-Darren committed:
Testament.java: Modified dialogue

Change dialogue of ToDo to contain more information

O-Darren committed:
Testament.java: Implement deadlines

Implement support for deadlines

O-Darren committed:
Testament.java: remove original tasks

Testament no longer stores any words from user as a task

O-Darren committed:
Task.java: create factory method

create a factory method that takes in a string and returns different types of tasks depending on the first word of the string

O-Darren committed:
Testament.java: Change task generation

Use factory method to create new tasks

O-Darren committed:
Implement Events class

O-Darren committed:
Task.java: add Events to factory method

Factory method creates events with the correct identifier string

O-Darren committed:
Fix runtest.bat

Change name of main file in runtest.bat

O-Darren committed:
implement unit tests

write basic unit tests that check functionality

O-Darren committed:
Implement NotInTaskListException

Implement exception that represents a task not being in the tasklist

O-Darren committed:
NotInTaskException.java: Change error message

O-Darren committed:
TaskList.java: change methods to throw NotInTaskException

mark, unMark, and getTask now throw NotInTaskException

O-Darren committed:
Testament.java: implement NotIntTaskException

add try catch blocks to deal with NotInTaskExceptions being thrown

O-Darren committed:
Implement CommandNotRecognisedException

O-Darren committed:
Task.java: Throw CommandNotRecognisedException

Factory method of throws CommandNotRecognisedException when it cannot identify the type of task

O-Darren committed:
Testament.java: catch CommandNotRecognisedException

Add try catch block to handle CommandNotRecognisedException

O-Darren committed:
Implement TaskDescriptionEmptyException

Implement an exception that represents the task description being empty. The string in the constructor represents the type of task

O-Darren committed:
Task.java: Constructor throws TaskDescriptionEmptyException

Constructor will throw TaskDescriptionEmptyException if user does not provide details for the task

O-Darren committed:
Implement TestamentException.java

Implement exception that represents all exceptions related to Testament chatbot

O-Darren committed:
Change Exceptions

made all existing exemptions related to Testament chatbot extend from TestamentException

O-Darren committed:
Testament.java: Change try catch block to catch TestamentException

O-Darren committed:
Implement DeadlineNoDateException

Implement exception that represents user not providing date for deadline

O-Darren committed:
Task.java: Implement support for DeadlineNoDateException

Added try catch block that throws DeadlineNoDateException if date is missing from deadline user input

O-Darren committed:
Implement EventNoTimeException

Implement exception that represents user not providing time for event

O-Darren committed:
Task.java: Implement EventNoTimeException support

Task.java now throws EventNoTimeException when user does not provide time for event

O-Darren committed:
Testament.java: Added check for mark and unmark

mark and unmark now check if the user specifies a task to be marked/unmarked

O-Darren committed:
Testament.java: handle non number being inputted for task

Now prints message telling user that tasks should be specified by their task number

O-Darren committed:
TaskList.java: Add deleteTask function

implement a function that deletes a task from the taskList

O-Darren committed:
TaskList.java: bugfix deleteTask function

O-Darren committed:
Testament.java: Implement delete

User can now delete tasks by specifying delete (tasknumber)

O-Darren committed:
Create document for Testament to store taskList data

O-Darren committed:
Implement methods that convert tasks to Strings

O-Darren committed:
Implement memory system in testament

O-Darren committed:
Storage.java: File creation

Create memory.txt if it is missing

O-Darren committed:
Merge branch 'branch-Level-7'

O-Darren committed:
Deadline.java: Change date storage to localDate

O-Darren committed:
Events.java: Change date storage to localDate

Change task.java parsing in factory method to support the change

O-Darren committed:
Testament.java: handle DateTimeParseException

O-Darren committed:
Events.java: Remove debug print statements

O-Darren committed:
Merge branch 'branch-Level-8'

O-Darren committed:
Implement Ui.java

O-Darren committed:
CommandNotRecognisedException.java: New constructor

Allow developer to send custom exception messages

O-Darren committed:
Implement Enumerations.java

O-Darren committed:
Implement Parser.java

Modify Testament.java and TaskList.java to work with Parser.java

O-Darren committed:
Merge branch 'A-MoreOOP'

O-Darren committed:
Implement Packages

O-Darren committed:
Merge branch 'A-Packages'

O-Darren committed:
Merge branch 'add-gradle-support'

O-Darren committed:
Add gradle support

O-Darren committed:
Merge branch 'branch-A-Gradle'

O-Darren committed:
Implement TaskTest.java

O-Darren committed:
Implement EnumerationsTest.java

O-Darren committed:
Merge branch 'branch-A-JUnit'

O-Darren committed:
Change main file path in build.gradle

O-Darren committed:
Testament.java: change file path

O-Darren committed:
Merge branch 'branch-A-Jar'

O-Darren committed:
Testament.java: Add Javadoc comments

O-Darren committed:
Ui.java: Add Javadoc comments

O-Darren committed:
Add periods to existing Javadoc comments

O-Darren committed:
TaskList.java: add Javadoc comments

O-Darren committed:
Task.java: add Javadoc comments

O-Darren committed:
ToDo.java: add Javadoc comments

O-Darren committed:
Events.java: add Javadoc comments

O-Darren committed:
Events.java: add periods to Javadoc comments

O-Darren committed:
Deadline.java: add Javadoc comments

O-Darren committed:
Storage.java: add Javadoc comments

O-Darren committed:
Storage.java: add periods to Javadoc comments

O-Darren committed:
Enumerations.java: add Javadoc comments

O-Darren committed:
Parser.java: add Javadoc comments

O-Darren committed:
TestamentException.java: add Javadoc comments

O-Darren committed:
TaskDescriptionEmptyException.java: add Javadoc comments

O-Darren committed:
NotInTaskListException.java: add Javadoc comments

O-Darren committed:
CommandNotRecognisedException.java: add Javadoc comments

O-Darren committed:
DeadlineNoDateException.java: add Javadoc comments

O-Darren committed:
EventNoTimeException.java: add Javadoc comments

O-Darren committed:
Merge branch 'branch-A-JavaDoc'

O-Darren committed:
Ui.java: adhere to coding standard

O-Darren committed:
Parser.java: adhere to coding standard

O-Darren committed:
Storage.java: adhere to coding standard

O-Darren committed:
Task.java: adhere to coding standard

O-Darren committed:
TaskList.java: adhere to coding standard

O-Darren committed:
Testament.java: adhere to coding standard

O-Darren committed:
Merge branch 'branch-A-CodingStandard'

O-Darren committed:
Enumerations.java: add FIND constant

O-Darren committed:
TaskList.java: Implement findDescription function

O-Darren committed:
Ui.java: implement find function

O-Darren committed:
Parser.java: Add support for finding tasks

O-Darren committed:
Merge branch 'branch-Level-9'

casperngeen commented on src/main/java/Parser/Enumerations.java:
```diff
@@ -0,0 +1,36 @@
+
+    private final String text;
+    private Enumerations(String s) {
+        text = s;
```
```suggestion
        this.text = s;
```

O-Darren replied:
Noted! Will change 😄

casperngeen commented on src/main/java/Parser/Parser.java:
```diff
@@ -0,0 +1,157 @@
+     * @return Task number.
+     * @throws TestamentException thrown if user did not enter task number in input, or if format is wrong.
+     */
+    private int validateTaskNumber(String[] splitUserInput) throws TestamentException {
```
```splitUserInput``` could be renamed into a noun, eg. ```userInputs```

casperngeen commented on src/main/java/Tasks/Task.java:
```diff
@@ -0,0 +1,127 @@
+ * Represents a task stored in the testament chatbot.
+ */
+public abstract class Task {
+    protected String taskname;
```
```suggestion
    protected String taskName;
```

casperngeen commented on src/main/java/Tasks/Task.java:
```diff
@@ -0,0 +1,127 @@
+     *
+     * @param s Description of task.
+     */
+    public Task(String s) {
+        taskname = s;
+        isDone = false;
```
```suggestion
    public Task(String description) {
        this.taskName = description;
        this.isDone = false;
```

casperngeen reviewed (commented):
LGTM :smile: 

I just added a few suggestions on variable/argument naming, and using this.var for your fields.

O-Darren committed:
Rename packages to start with small letter

O-Darren committed:
Implement checkstyle, make changes to style

O-Darren committed:
Merge branch 'branch-A-CheckStyle'

O-Darren committed:
Update build.gradle to support JavaFX

O-Darren committed:
Testament.java: Change constructor

Constructor does not take parameters

O-Darren committed:
Create files for gui

O-Darren committed:
Modify build.gradle main class to launcher

O-Darren committed:
Modify classes to work with gui

O-Darren committed:
Testament.java: Remove run method and modify main method

O-Darren committed:
Merge branch 'branch-Level-10'

O-Darren committed:
Fix style errors

O-Darren committed:
Testament.java: Change file path

O-Darren committed:
Storage.java: Create parent directory of memory

O-Darren committed:
Modify build.gradle to enable assertions

O-Darren committed:
Storage.java: Add assertions

O-Darren committed:
Merge pull request #1 from O-Darren/branch-A-Assertions

Branch a assertions

O-Darren committed:
Parser.java: Refactor parse method

The parse method is extremely long, with many cases.

Extracting each case into its own method improves code quality.

O-Darren committed:
Ui.java: Change schedule message

The schedule message only prints the schedule, with no message
from the chatbot

Adding a message gives the chatbot more personality

O-Darren committed:
Storage.java: Refactor load method

The load method has three if cases.

Extracting the code in each case into its own method improves
the quality of the code

O-Darren committed:
Merge branch 'master' into branch-A-CodeQuality

O-Darren committed:
Merge pull request #2 from O-Darren/branch-A-CodeQuality

Branch a code quality

O-Darren committed:
Merge branch 'master' of https://github.com/O-Darren/ip

O-Darren committed:
Enumerations.java: Add case for archive

O-Darren committed:
Ui.java: Add archive message

O-Darren committed:
TaskList.java: Add deleteTaskList method

O-Darren committed:
Storage.java: Add writeArchive and loadArchive

Save and load methods have code extracted to avoid duplicate code
in writeArchive and loadArchive methods

O-Darren committed:
Parser.java: Add archive command

O-Darren committed:
Enumerations.java: Add restore case

O-Darren committed:
Ui.java: Add retrieveArchive message

O-Darren committed:
Parser.java: Handle retrieveArchive

O-Darren committed:
Merge branch 'C-Archive'

O-Darren committed:
Parser.java: Extract default case

O-Darren committed:
Change profile image of Testament

O-Darren committed:
Add background image for Testament chatbot

O-Darren committed:
Change javadoc comments of toString methods

O-Darren committed:
Ui.java: Alternate message for empty schedule

O-Darren committed:
Remove instances of word Duke

O-Darren committed:
Fix style violations

O-Darren committed:
Ui.java: Change archive message

O-Darren committed:
Create user guide

O-Darren committed:
Bugfix welcome message not playing

O-Darren committed:
Relocate Ui.png

O-Darren committed:
Add title for jar file

O-Darren committed:
Add bye functionality and update user guide

O-Darren committed:
Task.java: Refactor factory method

Extract individual creation of each type of task

O-Darren committed:
Add credits to user guide
